## Cryptography GMI26H
# Dalarna University, 
# Autumn 2022
# Advanced Encyption Standard (AES)
### Deadline: 27 October, 2022, 23:59 CET time

### --------------------------------------------------------------------------------------------------------------------------------------------------------------

### Student name: Mikael Olsson
### Student ID: h20mikol
### University email address: h20mikol@du.se
### --------------------------------------------------------------------------------------------------------------------------------------------------------------

This document is a Jupyter notebook which will be used during the whole course. Students will run one lab every week as follows:

Week 3 of the course Lab 1

Week 4 of the course Lab 2

Week 5 of the course Lab 3

Week 6 of the course Lab 4, 5, and 6

## Cryptography Lab 1

## 
## 1. Key:

In the file **testKey**, a test key is provided. testKey is in plain text but represents bytes. Since AES-256 uses a key that is 256 bit long it corresponds to a key that is 256/8 bytes long. That is, the key is 32 bytes long, where each byte is encoded as two characters in hexadecimals. That is, the 256-bit key is represented as a string of length 64 hexamdecimal characters. <font color=red>We want the key to be transformed into a list of hexadecimal integers.</font>

**Method: getKey(filename)**

Create the method that does the following:

**Method name:** getKey()

**Input:** String filename

**Output:** List of integers

**Hint:**
The following Python methods may help you to build the getKey():

• hex()

• open()

• file.read()

• string.decode()

• map()

• ord()

In [2809]:
def getKey(testKey):
    file  = open(testKey,"r") # opens the file
    key = file.read() # sets key to be equal to the key stored in the file
    
    hexa_list = [] 
    for i in range(int(len(key)/2)): # i iterates from 0 to 31
        first_char = key[i*2] # sets the first character to be equal to i * 2 since there are 64 characters which will get every other character
        second_char = key[(i *2) + 1] # sets the second character to be (i * 2) + 1 since we want the character to the right of the previous one
        
        char = first_char + second_char # put the 2 characters together
        
        hexa_list.append(char) # add the hexadecimal to the list

    int_list = []
    for i in hexa_list: # iterate over the created list
        int_list.append(int(i,base=16)) # convert the hexadecimal to integer    

    return(int_list) # return the converted list


The following code can be used to test your method implementation. The list after "Output:" is what the "print(key)" command should print out.

In [2810]:
key = getKey("testKey")
print(key)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


**Output:**

    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

## 2. Block

In the file **testBlock**, a test block, which is a plain text but represents bytes, is provided. Since all versions of AES operate on blocks with the fix size of 128 bits, it corresponds to a block that is 128/8 bytes long. That is, the block is 16 bytes long, where each byte is encoded as two characters in hexadecimal. The 128-bit block is thus represented as a string of length 32.

In reality blocks have the format of a 4x4 matrix. In Python, matrices are typically lists. The first 4 values can be represented as the first column of the matrix, the next 4 as the second column, and so on. You can decide your format of the block yourself.

**Method: getBlock(filename)**

Create method that does the following:

**Method name:** getBlock()

**Input:** String filename

**Output:** List (or other format) of integers

**Hint:** The following Python methods may help you to build the method

hex()

open()

file.read()

string.decode()

map()

ord()

https://en.wikipedia.org/wiki/Advanced_Encryption_Standard#High-level_description_of_the_%20algorithm

In [2811]:
def getBlock(testBlock):
    file  = open(testBlock,"r") # opens the file
    block = file.read() # sets block to be equal to the key stored in the file
    
    hexa_list = []
    
    for i in range(int(len(block)/2)): # i iterates from 0 to 31
        first_char = block[i*2] # sets the first character to be equal to i * 2 since there are 64 characters which will get every other character
        second_char = block[(i *2) + 1] # sets the second character to be (i * 2) + 1 since we want the character to the right of the previous one
        
        char = first_char + second_char # put the 2 characters together
        
        hexa_list.append(char) # add the hexadecimal to the list

    int_list = []
    for i in hexa_list: # iterate over the created list
        int_list.append(int(i, base=16)) # convert the hexadecimal to integer

    return(int_list) # return the converted list
    

In [2812]:
block = getBlock("testBlock")
print(block)

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]


**Output:**

    [0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

## 3. sbox, sboxInv and rcon

From the following Wikipedia articles:

• https://en.wikipedia.org/wiki/Rijndael_S-box

• https://en.wikipedia.org/wiki/Rijndael_key_schedule

Create three Python code blocks, where each code block contains one of the three lists described in their respective Wikipedia articles. 

In [2813]:
hexa_sbox = ["63", "7C", "77", "7B", "F2", "6B", "6F", "C5", "30", "01", "67", "2B", "FE", "D7", "AB", "76",
            "CA", "82", "C9", "7D", "FA", "59", "47", "F0", "AD", "D4", "A2", "AF", "9C", "A4", "72", "C0",
            "B7", "FD", "93", "26", "36", "3F", "F7", "CC", "34", "A5", "E5", "F1", "71", "D8", "31", "15",
            "04", "C7", "23", "C3", "18", "96", "05", "9A", "07", "12", "80", "E2", "EB", "27", "B2", "75",
            "09", "83", "2C", "1A", "1B", "6E", "5A", "A0", "52", "3B", "D6", "B3", "29", "E3", "2F", "84",
            "53", "D1", "00", "ED", "20", "FC", "B1", "5B", "6A", "CB", "BE", "39", "4A", "4C", "58", "CF",
            "D0", "EF", "AA", "FB", "43", "4D", "33", "85", "45", "F9", "02", "7F", "50", "3C", "9F", "A8",
            "51", "A3", "40", "8F", "92", "9D", "38", "F5", "BC", "B6", "DA", "21", "10", "FF", "F3", "D2",
            "CD", "0C", "13", "EC", "5F", "97", "44", "17", "C4", "A7", "7E", "3D", "64", "5D", "19", "73",
            "60", "81", "4F", "DC", "22", "2A", "90", "88", "46", "EE", "B8", "14", "DE", "5E", "0B", "DB",
            "E0", "32", "3A", "0A", "49", "06", "24", "5C", "C2", "D3", "AC", "62", "91", "95", "E4", "79",
            "E7", "C8", "37", "6D", "8D", "D5", "4E", "A9", "6C", "56", "F4", "EA", "65", "7A", "AE", "08",
            "BA", "78", "25", "2E", "1C", "A6", "B4", "C6", "E8", "DD", "74", "1F", "4B", "BD", "8B", "8A",
            "70", "3E", "B5", "66", "48", "03", "F6", "0E", "61", "35", "57", "B9", "86", "C1", "1D", "9E",
            "E1", "F8", "98", "11", "69", "D9", "8E", "94", "9B", "1E", "87", "E9", "CE", "55", "28", "DF",
            "8C", "A1", "89", "0D", "BF", "E6", "42", "68", "41", "99", "2D", "0F", "B0", "54", "BB", "16"]
sbox = []

for i in hexa_sbox:
    sbox.append(int(i, 16))

hexa_sboxInv = ["52", "09", "6A", "D5", "30", "36", "A5", "38", "BF", "40", "A3", "9E", "81", "F3", "D7", "FB",
            "7C", "E3", "39", "82", "9B", "2F", "FF", "87", "34", "8E", "43", "44", "C4", "DE", "E9", "CB",
            "54", "7B", "94", "32", "A6", "C2", "23", "3D", "EE", "4C", "95", "0B", "42", "FA", "C3", "4E",
            "08", "2E", "A1", "66", "28", "D9", "24", "B2", "76", "5B", "A2", "49", "6D", "8B", "D1", "25",
            "72", "F8", "F6", "64", "86", "68", "98", "16", "D4", "A4", "5C", "CC", "5D", "65", "B6", "92",
            "6C", "70", "48", "50", "FD", "ED", "B9", "DA", "5E", "15", "46", "57", "A7", "8D", "9D", "84",
            "90", "D8", "AB", "00", "8C", "BC", "D3", "0A", "F7", "E4", "58", "05", "B8", "B3", "45", "06",
            "D0", "2C", "1E", "8F", "CA", "3F", "0F", "02", "C1", "AF", "BD", "03", "01", "13", "8A", "6B",
            "3A", "91", "11", "41", "4F", "67", "DC", "EA", "97", "F2", "CF", "CE", "F0", "B4", "E6", "73",
            "96", "AC", "74", "22", "E7", "AD", "35", "85", "E2", "F9", "37", "E8", "1C", "75", "DF", "6E",
            "47", "F1", "1A", "71", "1D", "29", "C5", "89", "6F", "B7", "62", "0E", "AA", "18", "BE", "1B",
            "FC", "56", "3E", "4B", "C6", "D2", "79", "20", "9A", "DB", "C0", "FE", "78", "CD", "5A", "F4",
            "1F", "DD", "A8", "33", "88", "07", "C7", "31", "B1", "12", "10", "59", "27", "80", "EC", "5F",
            "60", "51", "7F", "A9", "19", "B5", "4A", "0D", "2D", "E5", "7A", "9F", "93", "C9", "9C", "EF",
            "A0", "E0", "3B", "4D", "AE", "2A", "F5", "B0", "C8", "EB", "BB", "3C", "83", "53", "99", "61",
            "17", "2B", "04", "7E", "BA", "77", "D6", "26", "E1", "69", "14", "63", "55", "21", "0C", "7D"]
sboxInv = []

for i in hexa_sboxInv:
    sboxInv.append(int(i,16))


hexa_rcon = ["8d", "01", "02", "04", "08", "10", "20", "40", "80", "1b", "36", "6c", 
    "d8", "ab", "4d", "9a", "2f", "5e", "bc", "63", "c6", "97", "35", "6a", 
    "d4", "b3", "7d", "fa", "ef", "c5", "91", "39", "72", "e4", "d3", "bd", 
    "61", "c2", "9f", "25", "4a", "94", "33", "66", "cc", "83", "1d", "3a", 
    "74", "e8", "cb", "8d", "01", "02", "04", "08", "10", "20", "40", "80", 
    "1b", "36", "6c", "d8", "ab", "4d", "9a", "2f", "5e", "bc", "63", "c6", 
    "97", "35", "6a", "d4", "b3", "7d", "fa", "ef", "c5", "91", "39", "72", 
    "e4", "d3", "bd", "61", "c2", "9f", "25", "4a", "94", "33", "66", "cc", 
    "83", "1d", "3a", "74", "e8", "cb", "8d", "01", "02", "04", "08", "10", 
    "20", "40", "80", "1b", "36", "6c", "d8", "ab", "4d", "9a", "2f", "5e", 
    "bc", "63", "c6", "97", "35", "6a", "d4", "b3", "7d", "fa", "ef", "c5", 
    "91", "39", "72", "e4", "d3", "bd", "61", "c2", "9f", "25", "4a", "94", 
    "33", "66", "cc", "83", "1d", "3a", "74", "e8", "cb", "8d", "01", "02", 
    "04", "08", "10", "20", "40", "80", "1b", "36", "6c", "d8", "ab", "4d", 
    "9a", "2f", "5e", "bc", "63", "c6", "97", "35", "6a", "d4", "b3", "7d", 
    "fa", "ef", "c5", "91", "39", "72", "e4", "d3", "bd", "61", "c2", "9f", 
    "25", "4a", "94", "33", "66", "cc", "83", "1d", "3a", "74", "e8", "cb", 
    "8d", "01", "02", "04", "08", "10", "20", "40", "80", "1b", "36", "6c", 
    "d8", "ab", "4d", "9a", "2f", "5e", "bc", "63", "c6", "97", "35", "6a", 
    "d4", "b3", "7d", "fa", "ef", "c5", "91", "39", "72", "e4", "d3", "bd", 
    "61", "c2", "9f", "25", "4a", "94", "33", "66", "cc", "83", "1d", "3a", 
    "74", "e8", "cb", "8d"]
rcon = []

for i in hexa_rcon:
    rcon.append(int(i,16))


In [2814]:
#Check that you can run the following commands to see if the lists map integers correctly:
print(sbox[0])
print(sboxInv[0])
print(rcon[0])


99
82
141


**Output:**

    99 82 141

# Cryptography Lab 2

# 1. shiftRows

shiftRows is a static block-operator. Meaning that it performs the same action every time it is called. It simply shifts each row to the left the same number of steps as the current row number. See: https://en.wikipedia.org/wiki/Advanced_Encryption_Standard#The_ShiftRows_step

Depending on how you have structured your block the implementation for this will be different. A list can be used to represent the matrix, where the first 4 elements corresponds to the first column, the next 4 elements corresponds to the second column, and so on. You can use either matrices, lists of lists or just a normal list.

Since this method is static you can either just make a temporary block and move the elements around as described or you can loop over each row and perform the shift depending on the current row number.

Create the following two methods: One called shiftRows(block) and one called shiftRowsInv(block). These methods are extremely similar but the reverse shift just moves elements to the right instead of left.

In [2815]:
def shiftRows(block):
    temp = block[1]
    block[1] = block[5]
    block[5] = block[9]
    block[9] = block[13]
    block[13] = temp
    
    temp = block[2]
    block[2] = block[10]
    block[10] = temp
    temp = block[6]
    block[6] = block[14] 
    block[14] = temp
    
    temp = block[3]
    block[3] = block[15]
    block[15] = block[11]
    block[11] = block[7]
    block[7] = temp
    
    return block
    

In [2816]:
def shiftRowsInv(block):
    temp = block[13]
    block[13] = block[9]
    block[9] = block[5]
    block[5] = block[1]
    block[1] = temp

    temp = block[14]
    block[14] = block[6]
    block[6] = temp
    temp = block[2]
    block[2] = block[10]
    block[10] = temp
    
    temp = block[15]
    block[15] = block[3]
    block[3] = block[7]
    block[7] = block[11]
    block[11] = temp

    return block

In [2817]:
#Test the shiftRows() using this code

block = getBlock("testBlock")

shiftedBlock = shiftRows(block)

print(shiftedBlock)

[0, 85, 170, 255, 68, 153, 238, 51, 136, 221, 34, 119, 204, 17, 102, 187]


**Output:**

[0, 85, 170, 255, 68, 153, 238, 51, 136, 221, 34, 119, 204, 17, 102, 187]

In [2818]:
#Test the shiftRowsInv() using this Code:

block = getBlock("testBlock")

shiftedBlock = shiftRows(block)

unShiftedBlock = shiftRowsInv(shiftedBlock)

print(unShiftedBlock)

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]


**Output:**

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

# 2. mixColumn()

The mix of columns is a shuffle of each individual column in the block. This shuffling happens in a deterministic way depending on the values within the column. For consultation of the exact way of doing this, see the following Wikipedia article: https://en.wikipedia.org/wiki/Rijndael_mix_columns#MixColumns

You can either implement Galois multiplication and perform the mix directly. Or you use the pre-calculated lists found in the article. To do AES mixing of columns one only needs to multiply elements with 1, 2 or 3. And for the un-mixing 9, 11, 13 or 14.

**Create method mixColumn() as follows:**

This method should take one column perform the multiplications in accordance to the multiplication tables. Refer to the Wikipedia article.

Input: column

Output: column

Remember that in this context + means XOR. In Python, the hat-symbol (caret), ˆ, is used for XOR.

Also create the inverse method mixColumnInv().

In [2819]:
def mixColumn(column):
    # a, b and d all come from the wikipage for simplicity.
    a = column
    b = [2,1,1,3,3,2,1,1,1,3,2,1,1,1,3,2]
    d = [0,0,0,0]

    # this follows the formula set out on the wikipage
    d[0] = calculate(a[0] , b[0]) ^ calculate(a[1] , b[4]) ^ calculate(a[2] , b[8]) ^ calculate(a[3] , b[12])
    d[1] = calculate(a[0] , b[1]) ^ calculate(a[1] , b[5]) ^ calculate(a[2] , b[9]) ^ calculate(a[3] , b[13])
    d[2] = calculate(a[0] , b[2]) ^ calculate(a[1] , b[6]) ^ calculate(a[2] , b[10]) ^ calculate(a[3] , b[14])
    d[3] = calculate(a[0] , b[3]) ^ calculate(a[1] , b[7]) ^ calculate(a[2] , b[11]) ^ calculate(a[3] , b[15])

    return d


# I took this part from the slides, this is now multiplication in GF(2^8)
def calculate(x, y):
    if y == 1: # anything that you multiply by 1 is simply itself.
        return x
    shifted_x = (x << 1) & 0xff # if we're multiplying with something other than 1 then we want to shift it to the left by 1, as stated in the slides. the & 0xff part makes sure the bit values above the first 8 bits are ignored.
    if y == 2:
        if x < 128: # checks if the high bit is equal to 0 or 1. any value over 128 will mean that the high bit value is 1, and any value under 128 will mean that the high value is 0
            return shifted_x
        else:
            return shifted_x ^ 0x1b
    if y == 3:
        return calculate(x, 2) ^ x # as stated in the slides, multiplication done by 3 is actually just multiplication done by 2 XOR'd by the value a

In [2820]:
def mixColumnInv(column): # this part is pretty much the same as mixColumn so I will skip over this part
    d = column
    a = [14,9,13,11,11,14,9,13,13,11,14,9,9,13,11,14]
    b = [0,0,0,0]
 
    b[0] = calculateInv(d[0] , a[0]) ^ calculateInv(d[1] , a[4]) ^ calculateInv(d[2] , a[8]) ^ calculateInv(d[3] , a[12])
    b[1] = calculateInv(d[0] , a[1]) ^ calculateInv(d[1] , a[5]) ^ calculateInv(d[2] , a[9]) ^ calculateInv(d[3] , a[13])
    b[2] = calculateInv(d[0] , a[2]) ^ calculateInv(d[1] , a[6]) ^ calculateInv(d[2] , a[10]) ^ calculateInv(d[3] , a[14])
    b[3] = calculateInv(d[0] , a[3]) ^ calculateInv(d[1] , a[7]) ^ calculateInv(d[2] , a[11]) ^ calculateInv(d[3] , a[15])

    return b


# Since calculate is multiplication in GF(2^8) I use the calculate function in calculateInv to get to the values of y
def calculateInv(x,y):
    if y == 9: # (2*2*2)+1 = 9
        return calculate(calculate(calculate(x,2),2),2) ^ x # to get 9 I first took cal(cal(x,2),2),2) whuich is then the same as (2*2)*2 = 8 then the 8 ^ x = 9
    if y == 11: # (2*2*2)+2+1 = 11
        return calculate(calculate(calculate(x,2),2),2) ^ calculate(x,2) ^ x 
    if y == 13: # (2*2*2)+(2*2)+1 = 13
        return calculate(calculate(calculate(x,2),2),2) ^ calculate(calculate(x,2),2) ^ x
    if y == 14: # (2*2*2)+(2*2)+2 = 14
        return calculate(calculate(calculate(x,2),2),2) ^ calculate(calculate(x,2),2) ^ calculate(x,2)

Next create method mixColumns() that calls the mixColumn(col) for each column in your block and then returns the mixed block.

In [2821]:
def mixColumns(block):
    newBlock = []

    for i in range(4):
        currentColumn = []
        currentColumn.append(block[i*4])
        currentColumn.append(block[i*4+1])
        currentColumn.append(block[i*4+2])
        currentColumn.append(block[i*4+3])

        d = mixColumn(currentColumn)
        for i in d:
            newBlock.append(i)

        
    return newBlock

The corresponding inverse function should simply call the mixColumnInv() method for each column in your block and then return the mixed block.

In [2822]:

def mixColumnsInv(block):
    newBlock = []

    for i in range(4):
        currentColumn = []
        currentColumn.append(block[i*4])
        currentColumn.append(block[(i*4)+1])
        currentColumn.append(block[(i*4)+2])
        currentColumn.append(block[(i*4)+3])

        d = mixColumnInv(currentColumn)
        for i in d:
            newBlock.append(i)
            
    return newBlock

Use the following code to test your methods

In [2823]:
testBlock = getBlock("testBlock")

print(testBlock)

mixedBlock = mixColumns(testBlock)

print(mixedBlock)

unMixedBlock = mixColumnsInv(mixedBlock)

print(unMixedBlock)

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]
[34, 119, 0, 85, 102, 51, 68, 17, 170, 255, 136, 221, 238, 187, 204, 153]
[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]


**Output:**

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]
    
[34, 119, 0, 85, 102, 51, 68, 17, 170, 255, 136, 221, 238, 187, 204, 153]
    
[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]


# 3. subBytes

Create method subBytes() and apply the sBox to each element in the block and return it.

In [2824]:
def subBytes(block):
    for i in range(len(block)):
        block[i] = sbox[block[i]]
    return block

Create method subBytesInv() and apply sBoxInv() to each element in the block. And return it.

In [2825]:
def subBytesInv(block):
    for i in range(len(block)):
        block[i] = sboxInv[block[i]]
    return block

Use the follwing code to test your methods:

In [2826]:
testBlock = getBlock("testBlock")

print(testBlock)

substitutedBlock = subBytes(testBlock)

print(substitutedBlock)

unSubstitutedBlock = subBytesInv(substitutedBlock)

print(unSubstitutedBlock)

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]
[99, 130, 147, 195, 27, 252, 51, 245, 196, 238, 172, 234, 75, 193, 40, 22]
[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]


**Output:**
    
[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

[99, 130, 147, 195, 27, 252, 51, 245, 196, 238, 172, 234, 75, 193, 40, 22]

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

# Cryptography Lab 3

# 1. keyScheduleCore()

Read about this method in https://en.wikipedia.org/wiki/Rijndael_key_schedule#Key_schedule_core <br>
or in NIST document paragraph 5.2 (Key Expansion)<br>
**Input:** 4 byte word and Iteration Number

**Output:** 4 byte word

Explain what you do and why here in this cell (in English): The 2 first steps according to the wikipedia link says that you should do a left circular shift so the first for loop is shifting it all to the left and then setting the last value to what used to be the first. The second step is to use the apply the sbox to the "word", which is shown in the second loop. Then finally we xor the round constant (rcon) to add asymmetry, this makes it more secure. 


In [2827]:
def keyScheduleCore(word, iterate):
    
    # added simple comments to assist with the explanation above
    
    # first loop used to do a left circular shift

    word = rotWord(word)
    
    # second loop is to apply the sbox

    word = subBytes(word)
        
    
    # adding asymmetry using round constants
    word[0] ^= rcon[iterate]
    
    return word


# originally I included it n the keyschedulecore function but after realizing I had to reuse the rotword as detailed in the wikipage I decided to come back and make a separate function
def rotWord(word):
    temp = word[0]
    for i in range(3):
        word[i] = word[i+1]
    word[3] = temp    
    
    return word


In [2828]:
#Test your code
word = [1,2,3,4]


print(word)

newWord = keyScheduleCore(word,1)


print(newWord)

[1, 2, 3, 4]
[118, 123, 242, 124]


**Output:**

    [1, 2, 3, 4]

    [118, 123, 242, 124]

# 2. expandKey()

AES-operations are performed on blocks multiple times, in so-called rounds. Each round uses a specific round key. The round keys are created from the original key via an extension. The original key is 256 bit (32 bytes) but needs to be expanded to a 240 byte key. This extension is done using Rijndael key schedule, see https://en.wikipedia.org/wiki/Rijndael_key_schedule#Key_schedule_description. <br>
Follow the algorithmic description in the Wikipedia page and create a method called expandKey().

**Method name:** expandKey(key)

**Input:** 256 bit key

**Output:** extended 240 byte key

Explain what you do and why here in this cell (in English): First the key is divided up into the 4 byte words, every word therefore is composed of 4 separate bytes where every number is a byte. We then loop over and here I used the "NIST FIPS" document linked in the courseroom. Based on some research done online and in combination with the document I have created the next step, first we set it so that word is set to the previous word. We then have if and elif where we check for 1. if the length of the list w (list with all the words) is divisible by NK (the amount of words in the key at first) , if it is we do a left circular shift and then apply the sbox. The first element is then XOR'd by the round constant, which again is for security reasons to protect against slide attacks. 2. else if the length of the list w modulo Nk produces the rest 4 then we only apply the sbox.

After that we XOR every element in the previous word and then word -NK which is the Nk'th word away, as detaield in the chapter 5.2 text and the in the figure 11. After every word is done, for simpler code I decided to use lists within lists and the output does not expect that so there is an if loop where the lists inside of the list are removed so it's only integers in the list.

In [2829]:
# the whole code was partially inspirsed by this code https://github.com/boppreh/aes/blob/master/aes.py

def expandKey(key):
    Nr = 14 # got these from NIST FIPS document in the course room
    Nk = 8
    
    w = []  
    j = 0  
    
    while (j < Nk):
        w.append(0)
        w[j] = [key[4*j], key[4*j+1], key[4*j+2], key[4*j+3]]
        j += 1

    i = 1

    while len(w) < (Nr + 1) * 4:
            word = list(w[-1]) # previous word
            if len(w) % Nk == 0: # checks if the number of words is divisble by Nk             
                word = keyScheduleCore( word, i) # apply what is in the keyschedulecore
                i += 1
            elif len(w) % Nk == 4: # check if the number of words is divisble by Nk
                word = subBytes(word) # left cicular shift

            word = list(i^j for i, j in zip(word, w[-Nk])) # xor the previous word and the the -Nkth word

            w.append(word)
    
    expandedKey = []
    
    # removing the lists within lists
    for i in w: 
        for j in i:
            expandedKey.append(j)

    return expandedKey

In [2830]:
key = getKey("testKey")

print(key)

expandedKey = expandKey(key)

print(expandedKey)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 165, 115, 194, 159, 161, 118, 196, 152, 169, 127, 206, 147, 165, 114, 192, 156, 22, 81, 168, 205, 2, 68, 190, 218, 26, 93, 164, 193, 6, 64, 186, 222, 174, 135, 223, 240, 15, 241, 27, 104, 166, 142, 213, 251, 3, 252, 21, 103, 109, 225, 241, 72, 111, 165, 79, 146, 117, 248, 235, 83, 115, 184, 81, 141, 198, 86, 130, 127, 201, 167, 153, 23, 111, 41, 76, 236, 108, 213, 89, 139, 61, 226, 58, 117, 82, 71, 117, 231, 39, 191, 158, 180, 84, 7, 207, 57, 11, 220, 144, 95, 194, 123, 9, 72, 173, 82, 69, 164, 193, 135, 28, 47, 69, 245, 166, 96, 23, 178, 211, 135, 48, 13, 77, 51, 100, 10, 130, 10, 124, 207, 247, 28, 190, 180, 254, 84, 19, 230, 187, 240, 210, 97, 167, 223, 240, 26, 250, 254, 231, 168, 41, 121, 215, 165, 100, 74, 179, 175, 230, 64, 37, 65, 254, 113, 155

**Output:**

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 165, 115, 194, 159, 161, 118, 196, 152, 169, 127, 206, 147, 165, 114, 192, 156, 22, 81, 168, 205, 2, 68, 190, 218, 26, 93, 164, 193, 6, 64, 186, 222, 174, 135, 223, 240, 15, 241, 27, 104, 166, 142, 213, 251, 3, 252, 21, 103, 109, 225, 241, 72, 111, 165, 79, 146, 117, 248, 235, 83, 115, 184, 81, 141, 198, 86, 130, 127, 201, 167, 153, 23, 111, 41, 76, 236, 108, 213, 89, 139, 61, 226, 58, 117, 82, 71, 117, 231, 39, 191, 158, 180, 84, 7, 207, 57, 11, 220, 144, 95, 194, 123, 9, 72, 173, 82, 69, 164, 193, 135, 28, 47, 69, 245, 166, 96, 23, 178, 211, 135, 48, 13, 77, 51, 100, 10, 130, 10, 124, 207, 247, 28, 190, 180, 254, 84, 19, 230, 187, 240, 210, 97, 167, 223, 240, 26, 250, 254, 231, 168, 41, 121, 215, 165, 100, 74, 179, 175, 230, 64, 37, 65, 254, 113, 155, 245, 0, 37, 136, 19, 187, 213, 90, 114, 28, 10, 78, 90, 102, 153, 169, 242, 79, 224, 126, 87, 43, 170, 205, 248, 205, 234, 36, 252, 121, 204, 191, 9, 121, 233, 55, 26, 194, 60, 109, 104, 222, 54]

# 3. createRoundKey()

Method createRoundKey() samples the expanded key. In total there are 14 rounds in AES-256, and 14 round keys need to be extracted from the expanded key.

**Method:** createRoundKey(expandedKey, n)

**Input:** expandedKey (240 bytes) and round number n

**Output:** roundKey (16 bytes)

Explain what you do and why here in this cell (in English): I could not find much documentation anywhere but it was surprisingly simple when I understood what to do, so what we really are trying to do is simply take out the words from each round. That is then the round key, first I declare a list since I find it easier than working with an empty list. I then loop through every part of of expandedKey and take out the 16 elements of every round and add it to the roundKey list. There are a total of 14 round, + 1 for the round 0 which is the original key. If n = 0 and i = 0 n*16+i will equal 0 and if n = 0 and i = 1 then n*16+i will equal 1 and if n = 1 and i = 0 then n*16+i will equal 16. It's quite a simple loop.

In [2831]:
def createRoundKey(expandedKey, n):
    roundKey = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    for i in range(len(roundKey)):
        roundKey[i] = expandedKey[n*16+i]

    return roundKey

In [2832]:
key = getKey("testKey")
expandedKey = expandKey(key)
roundKey0 = createRoundKey(expandedKey,0)
roundKey7 = createRoundKey(expandedKey,7)
roundKey14 = createRoundKey(expandedKey,14)
print(expandedKey)
print(roundKey0)
print(roundKey7)
print(roundKey14)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 165, 115, 194, 159, 161, 118, 196, 152, 169, 127, 206, 147, 165, 114, 192, 156, 22, 81, 168, 205, 2, 68, 190, 218, 26, 93, 164, 193, 6, 64, 186, 222, 174, 135, 223, 240, 15, 241, 27, 104, 166, 142, 213, 251, 3, 252, 21, 103, 109, 225, 241, 72, 111, 165, 79, 146, 117, 248, 235, 83, 115, 184, 81, 141, 198, 86, 130, 127, 201, 167, 153, 23, 111, 41, 76, 236, 108, 213, 89, 139, 61, 226, 58, 117, 82, 71, 117, 231, 39, 191, 158, 180, 84, 7, 207, 57, 11, 220, 144, 95, 194, 123, 9, 72, 173, 82, 69, 164, 193, 135, 28, 47, 69, 245, 166, 96, 23, 178, 211, 135, 48, 13, 77, 51, 100, 10, 130, 10, 124, 207, 247, 28, 190, 180, 254, 84, 19, 230, 187, 240, 210, 97, 167, 223, 240, 26, 250, 254, 231, 168, 41, 121, 215, 165, 100, 74, 179, 175, 230, 64, 37, 65, 254, 113, 155, 245, 0, 37, 136, 19, 187, 213, 90, 114, 28, 10, 78, 90, 102, 153, 169, 242, 79, 224, 126, 87, 43, 170, 205, 248, 205,

**Output:**

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 165, 115, 194, 159, 161, 118, 196, 152, 169, 127, 206, 147, 165, 114, 192, 156, 22, 81, 168, 205, 2, 68, 190, 218, 26, 93, 164, 193, 6, 64, 186, 222, 174, 135, 223, 240, 15, 241, 27, 104, 166, 142, 213, 251, 3, 252, 21, 103, 109, 225, 241, 72, 111, 165, 79, 146, 117, 248, 235, 83, 115, 184, 81, 141, 198, 86, 130, 127, 201, 167, 153, 23, 111, 41, 76, 236, 108, 213, 89, 139, 61, 226, 58, 117, 82, 71, 117, 231, 39, 191, 158, 180, 84, 7, 207, 57, 11, 220, 144, 95, 194, 123, 9, 72, 173, 82, 69, 164, 193, 135, 28, 47, 69, 245, 166, 96, 23, 178, 211, 135, 48, 13, 77, 51, 100, 10, 130, 10, 124, 207, 247, 28, 190, 180, 254, 84, 19, 230, 187, 240, 210, 97, 167, 223, 240, 26, 250, 254, 231, 168, 41, 121, 215, 165, 100, 74, 179, 175, 230, 64, 37, 65, 254, 113, 155, 245, 0, 37, 136, 19, 187, 213, 90, 114, 28, 10, 78, 90, 102, 153, 169, 242, 79, 224, 126, 87, 43, 170, 205, 248, 205, 234, 36, 252, 121, 204, 191, 9, 121, 233, 55, 26, 194, 60, 109, 104, 222, 54]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]

[61, 226, 58, 117, 82, 71, 117, 231, 39, 191, 158, 180, 84, 7, 207, 57]

[36, 252, 121, 204, 191, 9, 121, 233, 55, 26, 194, 60, 109, 104, 222, 54]


# 4. addRoundKey()

addRoundKey() XORs each element of a roundkey with a block. <br>See: https://en.wikipedia.org/wiki/Advanced_Encryption_Standard#The_AddRoundKey_step

**Method:** addRoundKey(block,roundKey)

**Input:** block (16 bytes) and roundKey (16 bytes)

**Output:** block (16 bytes)

Explain what you do and why here in this cell (in English): Each element in the block should be XOR'd with corresponding element in the roundKey, again the code is quite simple. 


In [2833]:
def addRoundKey(block,roundKey):
    for i in range(len(block)):
        block[i] = block[i] ^ roundKey[i]
    
    return block

In [2834]:
key = getKey("testKey")
testBlock = getBlock("testBlock")
expandedKey = expandKey(key)
roundKey0 = createRoundKey(expandedKey,0)
addedRoundKeyToBlock = addRoundKey(testBlock,roundKey0)
print(addedRoundKeyToBlock)


[0, 16, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240]


**Output:**

[0, 16, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240]

# Cryptography Lab 4, 5, and 6

A large text file tWotW.txt is now available on Learn. Your task in Lab 6 is to read this file,Split it into blocks, Encrypt the blocks, and produce an encrypted version of the file. Then, read the encrypted file block by block and decrypt it. Upload the encrypted version of tWOW.txt to Learn, as well. <br>Labs 1-6 must be merged in one JNB file and that file must be compressed with the corresponding .pdf file and encryptedtWOW.txt (the encrypted version of tWOW.txt), that is, your final submission includes three files compressed into one .rar or .zip file:

1. AES labs 1-6 in one .ipynb file,
2. the .ipynb file converted into .pdf, and
3. encryptedtWOW.txt by you.

Add brief explanations regarding how you built each method, and why each method is needed (in English).

Sometimes it is good to define helper methods in addition to the main methods before the actual methods. You may add helper methods before the main methods in the same Python cell.

# Lab 4 
## Encryption

By now all essential parts of AES has been implemented. What is left to do is to call all the methods in the correct order to encrypt a block. We are going to follow the description provided here:

https://en.wikipedia.org/wiki/Advanced_Encryption_Standard#High-level_description_of_the_algorithm

More specifically, the encryption method should perform the following steps:


In [2835]:
# #**Initialize:**

# expandKey(key)

# #**0-th round:**

# createRoundKey(expandedKey,0)
# addRoundKey(block, roundKey)

# #**1-st to 13:th round:**

# createRoundKey(expandedKey,i)
# subBytes(block)
# shiftRows(block) 
# mixColumns(block)
# addRoundKey(block, roundKey)

# #**14:th round:**

# createRoundKey(expandedKey,14)
# subBytes(block)
# shiftRows(block)
# addRoundKey(block, roundKey)

# #**Return:** block


# encrypt(block, key)

**Method name:** encrypt(blocks, key)

**Input:** block and key

**Output:** encrypted block

Explain what you do and why here in this cell (in English): The key is first expanded, we "create" the roundkey by picking out the first words from the expanded key and then applying it to the block. Then we iterate from 1 - 13 and apply the same steps. First the i roundkey is "created", then we do a left circular shift of the block, then shift the rows, the mix the columns and finally apply the roundkey. Finally for the 14th round we "create" the 14th roundkey and we do the same as with 1 - 13 but we leave out the mix columns step. This is now encrypted.


In [2836]:
def encrypt(block,key):
    
    expandedKey = expandKey(key)

    roundKey = createRoundKey(expandedKey,0)
    block = addRoundKey(block,roundKey)
    
    for i in range(13):
        roundKey = createRoundKey(expandedKey,i +1)
        block = subBytes(block)
        block = shiftRows(block)
        block = mixColumns(block)
        block = addRoundKey(block, roundKey)
         
    roundKey = createRoundKey(expandedKey,14)
    block = subBytes(block)
    block = shiftRows(block)
    block = addRoundKey(block,roundKey)

    return block

In [2837]:
key = getKey("testKey")
block = getBlock("testBlock")
print(block)
encryptedBlock = encrypt(block,key)
print(encryptedBlock)

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]
[142, 162, 183, 202, 81, 103, 69, 191, 234, 252, 73, 144, 75, 73, 96, 137]


**Output:**

    [0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

    [142, 162, 183, 202, 81, 103, 69, 191, 234, 252, 73, 144, 75, 73, 96, 137]

# Lab 5

## Decryption

Again, see: https://en.wikipedia.org/wiki/Advanced_Encryption_Standard#High-level_description_of_the_algorithm


## decrypt(block, key)

**Input:** block and key

**Output:** decrypted block


In [2838]:
def decrypt(block,key):
    expandedKey = expandKey(key)
    
    roundKey = createRoundKey(expandedKey,14)
    block = addRoundKey(block, roundKey)
    block = shiftRowsInv(block)
    block = subBytesInv(block)
    
    for i in range(13,0,-1):
        roundKey = createRoundKey(expandedKey, i)
        block = addRoundKey(block,roundKey)
        block = mixColumnsInv(block)
        block = shiftRowsInv(block)
        block = subBytesInv(block)
        
    roundKey = createRoundKey(expandedKey,0)
    block = addRoundKey(block, roundKey)
    
    return block

Explain what you do and why here in this cell (in English): To decrypt we essentially just do the same stuff but reversed, with both of these functions we can now encrypt and decrypt keys. The for loop is set up to go in reverse so instead of going 1,2,3,4,....,13 it goes 13,12,11,10,....,1. According to the description of the algorithm it seemed like this is how was supposed to be.

In [2839]:
key = getKey("testKey")
block = getBlock("testBlock")
print(block)
encryptedBlock = encrypt(block,key)
print(encryptedBlock)
decryptedBlock = decrypt(encryptedBlock,key)
print(decryptedBlock)

[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]
[142, 162, 183, 202, 81, 103, 69, 191, 234, 252, 73, 144, 75, 73, 96, 137]
[0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]


**Output:**

    [0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

    [142, 162, 183, 202, 81, 103, 69, 191, 234, 252, 73, 144, 75, 73, 96, 137]

    [0, 17, 34, 51, 68, 85, 102, 119, 136, 153, 170, 187, 204, 221, 238, 255]

More specifically, the decryption method should perform the following steps:

In [2840]:
# #**Initialize:**

# expandedKey(key)

# #**14-th round:**

# createRoundKey(expandedKey,14)
# addRoundKey(block, roundKey)
# shiftRowsInv(block)
# subBytesInv(block)

# #**13-th to 1:st round:**

# createRoundKey(expandedKey,i)
# addRoundKey(block, roundKey)
# mixColumnsInv(block)
# shiftRowsInv(block)
# subBytesInv(block)

# #**0:th round:**

# createRoundKey(expandedKey,0)
# addRoundKey(block, roundKey)

# #**Return:** block

## This is my code for encrypting and decrypting and writing to file

In [2841]:
def encryptWotw(book,key):
    encryptedBook = []

    for i in range((len(book)//16)+ 1):
        temp = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        for j in range(16):
            try:
                temp[j] = int(ord(book[16*i+j]))
            except:
                break
        k = encrypt(temp, key)

        encryptedBook.append(k)
    
    return encryptedBook


def decryptWotw(key,fileName):
    decryptedBook = []
    w = open(fileName,'r',encoding="utf-8")
    w = w.read()
    w = w.split(',')
    
    
    for i in range((len(w)//16)):
        temp = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        for j in range(16):
            temp[j] = int(w[16*i+j])
            
        k = decrypt(temp, key)
        decryptedBook.append(k)
    
    return decryptedBook


def writeToFileEncrypted(bookList,fileName):
    f = open(fileName, "w",encoding="utf-8")
    
    cleanBook = []

    for i in range(len(bookList)):
        temp = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        for j in range(16):
            temp[j] = bookList[i][j]
            
        for i in temp:
            cleanBook.append(i)
        
    bookString = listToStringEncrypt(cleanBook)
    
    f.write(bookString)
    
def writeToFileDecrypted(bookList,fileName):
    f = open(fileName, "w",encoding="utf-8")
    
    cleanBook = []
    
    for i in range(len(bookList)):
        temp = []
        for j in range(16):
            if bookList[i][j] == 0:
                continue
            else:
                temp.append(0)
                temp[j] = chr(bookList[i][j])

        for i in temp:
            cleanBook.append(i)
    
    bookString = listToString(cleanBook)
    
    f.write(bookString)
            
def listToStringEncrypt(listBook):
    temp = ""
    for i in listBook:
        temp += str(i)
        temp += ','
    return temp

def listToString(listBook):
    temp = ""
    for i in listBook:
        temp += i
    return temp

In [2842]:
import time

originalBook = open("tWotW.txt","r")
originalBook = originalBook.read()
key = getKey("realKey") 

start = time.time()
encryptedBook = encryptWotw(originalBook,key)
writeToFileEncrypted(encryptedBook,"encryptedWotW.txt")
end = time.time()
encryptionTime = end - start


start = time.time()
decryptedBook = decryptWotw(key, "encryptedWotW.txt")
writeToFileDecrypted(decryptedBook,"decryptedWotW.txt")
end = time.time()
decryptionTime = end - start


print(f'Encryption time is {encryptionTime}')
print(f'Decryption time is {decryptionTime}')

Encryption time is 10.793421506881714
Decryption time is 28.206042289733887


# Lab 6

By now you should have implemented the methods encrypt() and decrypt(). Both take one block and one key as input and output a block. In the last lab you will apply your implementation to a larger file, analyse your results and prepare for your examination on the project.

A large text file tWotW.txt is now available on Learn. Your task is to read this file, split it into blocks,encrypt the blocks, and produce an encrypted version of the file. Read the encrypted file block by block and decrypt it.

## Key

### Question 1: What is the encryption key are you using (in hex format)?

Question 1 answer: 2b66eefc3568b5de89b7abdaf9ccd45667b374e861ed7b6c2bc9fc007f9dc074 which is a randomly generated key from the internet.


### Question 2: What is the avalanche effect in cryptography? If one bit is changed in a key, what should happen to the ciphertext? This is a theoretical question, no need to demonstrate this in practice.

Question 2 answer: The avalanche effect states that a small change to the key should produce a big change in the ciphertext, the ciphertext should look very different. Similar to how an avalanche can start with something small but cause a huge avalanche



## Confusion and diffusion

See: https://en.wikipedia.org/wiki/Confusion_and_diffusion
### Question 3: What is confusion and how is it achieved in AES?

Question 3 answer: Confusion is used to obscure the connection between the key and the ciphertext, this is achieved by hiding the relationship between the key and the cyphertext. In AES this is achieved, if you replace the first number in testBlock from a 0 to 1, it completely changes the ciphertext (output from encrypt function). The main functions contributing to confusion in AES are addRoundKey amd subBytes, the round key uses the expanded key to obscure and subbytes uses an already calculated non-linear box to also achieve confusion.

### Question 4: What is diffusion and how is it achieved in AES?

Question 4 answer: The idea of diffusion is that if someone inputs the same number or gives a pattern of numbers that the pattern or the number should be clearly shown in the cipher. For example, if the input is all zeros. Then the output should for example be all ones, or if an input like "01020304" then an output shouldn't be "10203040". Diffusion would aim at destroying these patterns, the functions that are used to provide full diffusion are shiftRows and mixColumns. These 2 work together to switch the places of the values in the input key to make them appear in different places in the output ciphertext.

## Mode of operation

See: https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Common_modes

### Question 5: You are free to apply any mode of operation described in the article above. ECB is the easiest to implement, but it has some weaknesses. What are these weaknesses?

Question 5 answer: Lack of diffusion, since it takes blocks and encrypts them individually the ciphertext produced can have a pretty clear pattern. After some googling I found out that the 2013 adobe password leak was partially due to ECB being used for encryption. ECB can also make protocols more susceptible to replay attacks but it seems like the main problem is the lack of diffusion.


source for the adobe leak: https://nakedsecurity.sophos.com/2013/11/04/anatomy-of-a-password-disaster-adobes-giant-sized-cryptographic-blunder/

### Question 6: A stronger mode from a security point-of-view is CBC. It is not much harder to implement but much stronger. What are the advantages and disadvantages of CBC compared to ECB?

Question 6 answer: CBC works by XOR'ing the previous ciphertext block with the next plaintext block. This then provides diffusion by obscuring the link between data, in ECB 2 identical ciphertext blocks would produce the exact same plaintext blocks but in CBC that would not happen.

### Question 7: What mode of operation are you using? Describe how it works.

Question 7 answer: I chose to use ECB. ECB divides the text into blocks and encrypts each block separately, this then produces blocks of ciphertext.



## Performance

### Question 8: How many blocks does the file tWotW.txt consist of?

Question 8 answer: 21501


### Question 9: How long does it take to read the file, encrypt it and store (write to disk) the encrypted file? You can import and use the time class in Python for this. Use any key you wish.

Question 9 answer:



### Question 10: How long does it take to read the encrypted file, decrypt it and store the plain text file?

Question 10 answer:



### Question 11: Do the encryption and decryption times differ? If so, why?

Question 11 answer:



### Question 12: Give suggestions regarding how you could improve the performance of your code.

Question 12 answer:



### Question 13: Try to compress (.zip or other compression) both the plain text file and the encrypted file. Are there any differences in the size of these files? If so why? Is it better to compress first and encrypt later or encrypt first and compress later? Motivate your answer!

Question 13 answer:



### Notes:

During your presentation we will look at your code together and we will ask questions regarding your implementation. You must be able to explain how all parts of your code works. Feel free to add comments to the code and text cells to help your explanation.

If you get help from others you **MUST** comment this in your code. You should comment if you have discussed your code or collaborated with fellow students and comment this for every method and give credit to the right person. The same goes if you have looked at code from the internet. Provide links.

Upload the encrypted version encryptedtWOW.txt of the original file tWOW.txt to Learn, as well. Make sure that the decrypted file is identical to the original file tWOW.txt.